### PR TITLE
Blog post - Monitoring Canary Releases with Hawkular APM

### PR DESCRIPTION
**Pull request by jpkrohling**

### CODE REVIEW

**Comment by jpkrohling** [comment]
@theute , @objectiser , could you please proof-read/test/review/merge?

File: src/main/jbake/content/blog/2017/02/13/monitoring-canary-deployments.adoc [hunk 1]
```diff
@@ -0,0 +1,249 @@
...
+
+Creating an order for account=steve item=car quantity=2
+{
+  "accountId" : "steve",
+  "itemId" : "car",
+  "quantity" : 2,
+  "id" : "7f095b1e-94a2-4398-8480-01aaea43e021"
+}
+...
+...
```
Comment (objectiser) [inline src/main/jbake/content/blog/2017/02/13/monitoring-canary-deployments.adoc]: May be removed these blank lines

File: src/main/jbake/content/blog/2017/02/13/monitoring-canary-deployments.adoc [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+```
+
+Hawkular APM should be located under the service name `hawkular-apm`, so, the URL
+be similar to this: `https://hawkular-apm-order-manager.192.168.2.107.xip.io`. Note
```
Comment (objectiser) [inline src/main/jbake/content/blog/2017/02/13/monitoring-canary-deployments.adoc]: This URL is for the order manager service - did you mean to refer to the APM server, to load the APM UI?

File: src/main/jbake/content/blog/2017/02/13/monitoring-canary-deployments.adoc [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+endif::[]
+
+Our Account Manager microservice is a Vert.x Verticle, and reacts on messages sent
+to `AccountManager.getAccount`. As we are using a clustered event bus, our new
+_account-manager-blue_ service will join the cluster and start replying to those
+messages. The Vert.x clustered bus will take care of load balancing the processing
+of the messages for us.
+
+Because of that, we should already have some data being captured and reported to
+our Hawkular APM server. On the _Services_ screen, add the new service _order-manager.account-manager_,
```
Comment (objectiser) [inline src/main/jbake/content/blog/2017/02/13/monitoring-canary-deployments.adoc]: Shouldn't this be account-manager-blue for the service name and version?

File: src/main/jbake/content/blog/2017/02/13/monitoring-canary-deployments.adoc [hunk 4]
```diff
@@ -0,0 +1,249 @@
...
+
+ifndef::env-github[]
+image::/img/blog/2017/2017-02-13-account-manager-blue-added.png[Hawkular APM Services Tab with Blue and Green services]
+endif::[]
+ifdef::env-github[]
+image::../../../../../assets/img/blog/2017/2017-02-13-account-manager-blue-added.png[Hawkular APM Services Tab with Blue and Green services]
+endif::[]
+
+We should see the bar for our _blue_ deployment (orange bar on the image) to be
+around 20% of the size of the old version (blue bar on the image). As the number
```
Comment (objectiser) [inline src/main/jbake/content/blog/2017/02/13/monitoring-canary-deployments.adoc]: Should be 25% - i.e. 20% are new version, 80% are old version, so new version is a quarter of the old version.

File: src/main/jbake/content/blog/2017/02/13/monitoring-canary-deployments.adoc [hunk 5]
```diff
@@ -0,0 +1,249 @@
...
+
+ifndef::env-github[]
+image::/img/blog/2017/2017-02-13-final.png[Final Result]
+endif::[]
+ifdef::env-github[]
+image::../../../../../assets/img/blog/2017/2017-02-13-final.png[Final Result]
+endif::[]
+
+On a real world scenario, the manual steps we did would be automated, potentially
+making your Continuous Deployment scripts to read data from Hawkular APM in order
```
Comment (objectiser) [inline src/main/jbake/content/blog/2017/02/13/monitoring-canary-deployments.adoc]: Remove 'to', so just read "scripts read data from Hawkular APM..."

**Review by objectiser — COMMENTED** [review]
Very good - just a couple of areas to change.

Only other point is the load balancing in OpenShift - although this works with vertx event bus, wondering if the same approach would work with REST endpoints - i.e. can a service/route??? be defined across multiple service deployments? (i.e. account-manager and account-manager-blue)?

**Comment by pilhuhn** [comment]
Put a detail of the last image (e.g. top chart with 3 color parts) at the top of the post below the heading as a teaser for more details.
Or a diagram that shows the architecture you are working on

File: src/main/jbake/content/blog/2017/02/13/monitoring-canary-deployments.adoc [hunk 6]
```diff
@@ -0,0 +1,249 @@
...
+...
+...
+```
+
+Hawkular APM should be located under the service name `hawkular-apm`, so, the URL
+be similar to this: `https://hawkular-apm-order-manager.192.168.2.107.xip.io`. Note
+that the final part of the hostname might be different.
+
+After a few minutes with our `genorders.sh` script running, we should be able to
+see a screen similar to this:
```
Comment (pilhuhn) [inline src/main/jbake/content/blog/2017/02/13/monitoring-canary-deployments.adoc]: You are using "similar to this" 3 times in a row. Try to change that

File: src/main/jbake/content/blog/2017/02/13/monitoring-canary-deployments.adoc [hunk 7]
```diff
@@ -0,0 +1,249 @@
...
+oc start-build account-manager-blue --from-dir=. --follow
+oc new-app \
+  account-manager-blue \
+  -l app=account-manager-blue \
+  HAWKULAR_APM_URI='http://hawkular-apm' \
+  HAWKULAR_APM_USERNAME='admin' \
+  HAWKULAR_APM_PASSWORD='password'
+```
+
+An output similar to this is expected:
```
Comment (pilhuhn) [inline src/main/jbake/content/blog/2017/02/13/monitoring-canary-deployments.adoc]: similar to this -> like the following

File: src/main/jbake/content/blog/2017/02/13/monitoring-canary-deployments.adoc [hunk 8]
```diff
@@ -0,0 +1,249 @@
...
+      you will need to create a service with 'expose dc/account-manager-blue --port=[port]' later
+
+--> Creating resources with label app=account-manager-blue ...
+    deploymentconfig "account-manager-blue" created
+--> Success
+    Run 'oc status' to view your app.
+```
+
+We should also have a new service displayed on OpenShift console. Note the new
+_account-manager-blue_ deployment, with 1 pod.
```
Comment (pilhuhn) [inline src/main/jbake/content/blog/2017/02/13/monitoring-canary-deployments.adoc:140]: You could put an orange circle around the -blue one on the image to make it easier to spot it.

**Comment by pilhuhn** [comment]
I think you should create sections to logically split the post into something like
preparation
scaling up the blue canary
Summary in one picture (the "we can infer" part).

All in all a good post.

**Comment by theute** [comment]
I'm biased by the awesomeness of the feature + happiness to have instructions on how to demo this soon ;) 
:+1:

File: src/main/jbake/content/blog/2017/02/13/monitoring-canary-deployments.adoc (hunk 2 above)
Comment (jpkrohling) [reply src/main/jbake/content/blog/2017/02/13/monitoring-canary-deployments.adoc]: Good catch!

Comment (jpkrohling) [reply src/main/jbake/content/blog/2017/02/13/monitoring-canary-deployments.adoc]: No, wait, it's correct: the Ansible playbook deploys Hawkular APM on the Order Manager project, so, the URL is correct: `servicename-projectname.ip.xip.io` .

File: src/main/jbake/content/blog/2017/02/13/monitoring-canary-deployments.adoc (hunk 3 above)
Comment (jpkrohling) [inline src/main/jbake/content/blog/2017/02/13/monitoring-canary-deployments.adoc]: No, the service name on APM is the namespace+service from OpenShift, so, `order-manager.account-manager` is correct for the service name. We have the version on the next line, `order-manager.account-manager-1`, which might be repetitive, but those are two actual fields we have on the screen.

File: src/main/jbake/content/blog/2017/02/13/monitoring-canary-deployments.adoc (hunk 2 above)
Comment (objectiser) [reply src/main/jbake/content/blog/2017/02/13/monitoring-canary-deployments.adoc]: ah ok :)

File: src/main/jbake/content/blog/2017/02/13/monitoring-canary-deployments.adoc (hunk 3 above)
Comment (objectiser) [reply src/main/jbake/content/blog/2017/02/13/monitoring-canary-deployments.adoc]: Its the `-blue` part I was pointing out - this section is talking about the newer version being deployed which has a modified service name.

File: src/main/jbake/content/blog/2017/02/13/monitoring-canary-deployments.adoc [hunk 9]
```diff
@@ -66,16 +68,14 @@ Creating an order for account=steve item=car quantity=2
...
   "id" : "7f095b1e-94a2-4398-8480-01aaea43e021"
 }
-...
-...
 ```
 
 Hawkular APM should be located under the service name `hawkular-apm`, so, the URL
-be similar to this: `https://hawkular-apm-order-manager.192.168.2.107.xip.io`. Note
-that the final part of the hostname might be different.
+be follows the pattern: `https://hawkular-apm-order-manager.YOUR-IP.xip.io`. Note
```
Comment (objectiser) [inline src/main/jbake/content/blog/2017/02/13/monitoring-canary-deployments.adoc]: Remove the initial 'be'

File: src/main/jbake/content/blog/2017/02/13/monitoring-canary-deployments.adoc [hunk 10]
```diff
@@ -0,0 +1,258 @@
...
+:jbake-status: published
+:jbake-tags: blog, apm, opentracing, vertx, openshift
+
+In our last release, link:/blog/2017/02/04/hawkular-apm-service-deployments.html[we built a Services screen]
+on top of a feature added a couple of releases ago, the "service name" and "build
+stamps". These properties can be specified as an environment variable or system
+property, and are automatically detected when the monitored application runs on
+OpenShift.
+
+On this blog post, we'll show you how to use that screen to monitor a Canary release
```
Comment (objectiser) [inline src/main/jbake/content/blog/2017/02/13/monitoring-canary-deployments.adoc]: Another minor - "In this blog..."

**Review by objectiser — APPROVED** [review]

**Comment by jpkrohling** [comment]
I added the image changes suggested by @pilhuhn and fixed the last two items from @objectiser . If that looks good, I'll squash and update the PR.

**Comment by objectiser** [comment]
@jpkrohling LGTM :+1:

**Comment by jpkrohling** [comment]
PR updated.

**Comment by pilhuhn** [comment]
Thanks @jpkrohling !!

**Pull request merged**